src/charm.py

```diff
@@ -43,55 +46,94 @@ def new_func(self, *args, **kwargs):
             func(self, *args, **kwargs)
         except BlockedStatusError as e:
             self.unit.status = BlockedStatus(str(e))
+
     return new_func
 
 
 class AlertmanagerCharm(CharmBase):
-    _stored = StoredState()
-
     def __init__(self, *args):
-        log.debug('Initializing charm.')
+        log.debug("Initializing charm.")
         super().__init__(*args)
         self.framework.observe(self.on.config_changed, self.on_config_changed)
-        self.framework.observe(self.on['alerting'].relation_changed, self.on_alerting_changed)
+        self.framework.observe(
+            self.on["alerting"].relation_changed, self.on_alerting_changed
+        )
+        self.framework.observe(
+            self.on["alertmanager"].relation_changed, self.on_alertmanager_changed
+        )
+        self.framework.observe(
+            self.on["alertmanager"].relation_departed, self.on_alertmanager_departed
+        )
 
     @status_catcher
     def on_alerting_changed(self, event):
+        self.update_alerting(event.relation)
+
+    def update_alerting(self, relation):
         if self.unit.is_leader():
-            log.info('Setting relation data')
-            event.relation.data[self.app]['port'] = str(self.model.config['port'])
+            log.info("Setting relation data: port")
+            if str(self.model.config["port"]) != relation.data[self.app].get(
+                "port", None
+            ):
+                relation.data[self.app]["port"] = str(self.model.config["port"])
+
+            log.info("Setting relation data: addrs")
+            addrs = []
+            num_units = self.num_units()
+            for i in range(num_units):
+                addrs.append(
+                    UNIT_ADDRESS.format(
+                        self.meta.name, i, self.meta.name, self.model.name
+                    )
+                )
+            if addrs != json.loads(relation.data[self.app].get("addrs", "null")):
+                relation.data[self.app]["addrs"] = json.dumps(addrs)
+
+    @status_catcher
+    def on_alertmanager_changed(self, event):
+        if self.unit.is_leader():
+            self.configure()
+            for relation in self.model.relations["alerting"]:
+                self.update_alerting(relation)
+
+    @status_catcher
+    def on_alertmanager_departed(self, event):
+        if self.unit.is_leader():
```

Comment:
Nit Pick: Inverting this would lead to more flat (as opposed to nested) code flow. So for instance if the check was `if not self.unit.is_leader()` then return. I do see you seem to have a preference for the other way round here and elsewhere in your pull request.

Reply:
FWIW the nested approach is more python-ish and the non-nested one is more Go-ish :-)

Reply:
As this is a purely preference decision. I find code with fewer "not"s to be more satisfying.